File: MovieAppMVVMRetrofitJava/app/src/main/java/com/masterclass/movieappmvvmretrofitjava/repositories/MovieRepository.java
package com.masterclass.movieappmvvmretrofitjava.repositories;

import android.app.Application;

import androidx.lifecycle.MutableLiveData;

import com.masterclass.movieappmvvmretrofitjava.R;
import com.masterclass.movieappmvvmretrofitjava.api.MovieApiService;
import com.masterclass.movieappmvvmretrofitjava.api.RetrofitInstance;
import com.masterclass.movieappmvvmretrofitjava.models.Movie;
import com.masterclass.movieappmvvmretrofitjava.models.Result;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MovieRepository {
    //A repository is used for abstracting all the data sources details and provides
    //a clean API for the ViewModel to fetch and manage data

    private ArrayList<Movie> movies = new ArrayList<>();
    private MutableLiveData<List<Movie>> mutableLiveData = new MutableLiveData<List<Movie>>();
    private Application application;

    public MovieRepository(Application application) {
        this.application = application;
    }

    public MutableLiveData<List<Movie>> getMutableLiveData(){
        MovieApiService movieApiService = RetrofitInstance.getService();

        Call<Result> call = movieApiService.
                getPopularMovies(application.getApplicationContext().getString(R.string.api_key));

        //enqueue executes the network request in the background and handles the response in the Main UI Thread
        //Unlike the execute which handles everything in the Main UI Thread
        call.enqueue(new Callback<Result>() {
            @Override
            public void onResponse(Call<Result> call, Response<Result> response) {
                Result result = response.body();
                if (result != null && result.getResults() != null){
                    movies = (ArrayList<Movie>) result.getResults();
                    mutableLiveData.setValue(movies);
                }
            }

            @Override
            public void onFailure(Call<Result> call, Throwable t) {

            }
        });

        return mutableLiveData;
    }
}

File: MovieAppMVVMRetrofitKotlin/app/src/main/java/com/masterclass/movieappmvvmretrofitkotlin/repositories/MovieRepository.kt
package com.masterclass.movieappmvvmretrofitkotlin.repositories

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.masterclass.movieappmvvmretrofitkotlin.R
import com.masterclass.movieappmvvmretrofitkotlin.api.RetrofitInstance
import com.masterclass.movieappmvvmretrofitkotlin.models.Movie
import com.masterclass.movieappmvvmretrofitkotlin.models.MoviesResult
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MovieRepository(private val application: Application) {
    private var movies = ArrayList<Movie>()
    private val mutableLiveData = MutableLiveData<List<Movie>>()

    fun getMutableLiveData(): MutableLiveData<List<Movie>> {
        val movieApiService = RetrofitInstance.getService()
        val apiKey = application.getString(R.string.api_key)

        val call = movieApiService.getPopularMovies(apiKey)

        call.enqueue(object : Callback<MoviesResult> {
            override fun onResponse(call: Call<MoviesResult>, response: Response<MoviesResult>) {
                val result = response.body()
                if (result?.results != null) {
                    movies.clear()
                    mutableLiveData.value = (result.results ?: emptyList()) as List<Movie>?
                }
            }

            override fun onFailure(call: Call<MoviesResult>, t: Throwable) {
                // Handle failure
            }
        })

        return mutableLiveData
    }
}
